market_maker_keeper/erisx_market_maker_keeper.py

```diff
@@ -376,5 +379,49 @@ def place_order_function(new_order_to_be_placed):
             else:
                 logging.info(f"New {side} Order below size minimum of {min_amount}. Order of amount {amount} ignored.")
 
+    def check_cancellations(self):
+        """
+        Check to see if ErisX has cancelled any of our orders.
+        If an order has been cancelled, remove it from the order book
+        """
+        with self._lock:
+            cancelled_orders = self.erisx_api.check_cancellations()
+
+            for order in cancelled_orders:
+                for index, existing_order in enumerate(self.orders):
+                    if order.order_id == existing_order.order_id:
+                        del self.orders[index]
+                        self.logger.info(f"Removed cancelled order: {order.order_id}")
+
+    def synchronize_orders(self):
+        bands = Bands.read(self.bands_config, self.spread_feed, self.control_feed, self.history)
+        order_book = self.order_book_manager.get_order_book()
+        target_price = self.price_feed.get_price()
+
+        # Remove any orders cancelled by ErisX
+        self.check_cancellations()
+
+        # Cancel orders
+        cancellable_orders = bands.cancellable_orders(our_buy_orders=self.our_buy_orders(order_book.orders),
+                                                      our_sell_orders=self.our_sell_orders(order_book.orders),
+                                                      target_price=target_price)
+        if len(cancellable_orders) > 0:
+            self.order_book_manager.cancel_orders(cancellable_orders)
+            return
+
+        # Do not place new orders if order book state is not confirmed
+        if order_book.orders_being_placed or order_book.orders_being_cancelled:
+            self.logger.debug("Order book is in progress, not placing new orders")
+            return
+
+        # Place new orders
+        self.place_orders(bands.new_orders(our_buy_orders=self.our_buy_orders(order_book.orders),
+                                            our_sell_orders=self.our_sell_orders(order_book.orders),
+                                            our_buy_balance=self.our_available_balance(order_book.balances,
+                                                                                        self.token_buy()),
+                                            our_sell_balance=self.our_available_balance(order_book.balances,
+                                                                                        self.token_sell()),
+                                            target_price=target_price)[0])
```

Comment:
What does the array returned by `new_orders` represent, and why are we only placing orders for the first element of it?

Reply:
`new_orders` represents the set of new orders that would comply with our bands config, and that haven't been placed already. We take the `[0]` index of `bands.new_orders` as it returns a Tuple, and we are specifically interested in the new_orders list that is the first returned. https://github.com/makerdao/market-maker-keeper/blob/master/market_maker_keeper/band.py#L325 The `missing_buy_amount` and `missing_sell_amount` Tuple returns appear to actually be dead code. Also, this logic should be consistent with other keepers.

Reply:
👍